### DISCUSSION TITLE
Proposal: Refcounted Double Linked Lists for version objects

### DISCUSSION THREAD

**Discussion by CharlesJQuarra** [skypjack/entt#761] [Ideas]
This might not be useful for everyone, but I thought to write it down in case it might have interesting applications

The problem with incremental counters for versions in `Entity` identifiers is that for long-running programs (like, if Entt/ECS is used to implement an online game server), they might wrap around and result in false alive tests.

The optimization opportunity is that many version numbers might have no entities pointing to them, and that all the functionality provided by versioning can presumably exist with a reduced set of Version identifiers. The only important invariants for versions is that:

1) they are well-ordered
2) that they remain alive as long as all past Entity references that point to it

These can be easily achieved by implementing the Version objects as DL lists with a refcount of Entity references. Whenever it is detected that no one holds an Entity reference point to a given Version object, the object is removed.

If one is confident the number of alive versions in a program will remain limited, one might replace a pointer to a version object as the version identifier, with an index in the actual array of pointers

Any feedback on the above is welcome.

**Comment by codylico** [comment]
Two thoughts come to mind:
- *Worst-case: one entity per version, all versions used up:* Is there a way to handle version identifier selection in this case?
- *New version selection:* How does this scheme select new version identifiers, given that a version object, once removed, "takes" its identifier with it? Do the pointers to object (for example, `Version* ptr;`) become the de-facto version identifier?

Forgive me if I'm missing something obvious.

**Comment by skypjack** [comment]
Converted the issue to a discussion since it seems better suited.
I'm not sure I fully get what you're proposing though. The basic idea would be to not embed the version into the entity identifier. Instead, you want to associate to every identifier a linked list of available versions that get recycled when the ref counter reaches 0 but that can also stay alive indefinitely, so that it would be normal to have multiple versions of the same entity around?

**Reply by CharlesJQuarra** [reply]
Every version id is replaced with a node pointer/index to a refcounted version entry. When a Entity id version must be promoted, the linked list is traversed a (hopefully small) number of nodes, ideally 1, but it could be more depending on the strategy used to avoid version creep. When a version refcount becomes zero, the version node is removed (with a linked list this takes O(1)). To avoid creep of version objects you might want to jump to elements in the linked list that have above average ref counts on their `next` neighbours, so that identifiers will tend to aggregate on fewer version nodes over time. 

On the end, the integrity of the version tests relies on version identifiers not getting recycled until no other reference points to it. This will add of course some constant costs to operating with Entities, but for the use case where this might be an issue, perhaps it is worth/unavoidable

**Reply by skypjack** [reply]
Indeed, it seems very expensive for the general case. However, I think it's somewhat viable already on top of `EnTT`.
Since we can force the version to use when deleting an entity, there is nothing that prevents users from implementing a similar model.
Have you already tried to do something similar? It doesn't seem trivial but I see how it can help in some cases.

**Reply by CharlesJQuarra** [reply]
I would like to try to write a benchmark for this. Besides eliciting general comments on the requirements and implementation ideas for long-term id recycling, I wanted to identify what changes I would need to do to make PoC for this.

Initially I am looking at the required specializations, as well as the methods [`entt::basic_registry::group_handler::release_entity`] and [`entt::basic_registry::clear`]. I am trying to understand the shortest way to test this out by overriding only the relevant parts to version id https://github.com/skypjack/entt/blob/0754f108c99ed2adce12b8d1bdf370169d103614/src/entt/entity/registry.hpp#L144
https://github.com/skypjack/entt/blob/0754f108c99ed2adce12b8d1bdf370169d103614/src/entt/entity/registry.hpp#L967

**Reply by skypjack** [reply]
🤔 what is needed for a PoC? That's a good question. I think it requires some work, since the design has never taken in consideration the possibility of having a version that isn't part of the identifier.
Probably you need to touch the `entt_traits` class so as to remove the version part of an identifier. Then, yeah, the work is mostly within the registry and likely the sparse set. The version isn't touched very often, so probably the number of functions to be updated is low but I don't remember them all by heart.

**Reply by CharlesJQuarra** [reply]
I added a draft PR: https://github.com/skypjack/entt/pull/767

Summarizing:

1) I still haven't figure it out the proper null/tombstone semantics for this case (should I make special non-zero value conversions to those global instances?)

2) my `entt_traits` specialization clashes with `entt_traits<Type, std::enable_if_t<std::is_class_v<Type>>>` specialization, but the good? news is that commenting that specialization does not break any existing tests

**Reply by skypjack** [reply]
The null identifier is such that both the entity and the version are 0xFF... so probably skipping it should be enough.
As for `entt_traits`, there is a test that uses the class version but it's not enabled by default. However, I'm pretty sure that dropping it will break existing code. In particular, that of a project on which I'm working. 😅

**Reply by skypjack** [reply]
Any performance results? From a quick glance, it seems way more expensive than the basic implementation.

**Reply by CharlesJQuarra** [reply]
I have some free time to try debugging the issues with the PR. However after looking at the code I'm still confused about the assertion at [sparse_set.hpp:518](https://github.com/skypjack/entt/blob/master/src/entt/entity/sparse_set.hpp#L518), it happens while trying to do `registry.insert(std::begin(entities), std::end(entities), 42)`. The error seems to suggest the entities cannot be distinguished from null or tombstone values. 

I was considering this approach: declare inside `entt_traits` two more `constexpr` calls called `construct_null` and `construct_tombstone` that resolve to `internal::entt_traits` implementation. For all types except those that need a long lived version, the specialization should resolve to the same `entt_traits::construct(entt_traits::entity_mask, entt_traits::version_mask)` call, but in my case, I could provide separate reserved instances so that they will not evaluate equal to normal entities

Is this semantically consistent with how you are using and defining nulls/tombstones?

**Reply by skypjack** [reply]
🤔 null entities and tombstones are interleaved with normal entities more often than not. For example, to track holes within stable pools, there is an implicit list in the packed array that exploits the _entity_ part of an identifier to store the _next position_ and set the tombstone as a version. Therefore, not sure if isolating the `construct` function addresses all uses.

**Reply by CharlesJQuarra** [reply]
so if I understand correctly, the null is a reserved entity section, and the tombstone is a reserved version section? I'm not clear how does the engine distinguish between nulls from tombstones when iterating on an array, if both resolve to the same `Entity` value

**Reply by skypjack** [reply]
There are both a null entity and a tombstone version around. `EnTT` doesn't really need to distinguish between them and actually it doesn't even want to. When the library encounters an identifier, it can check if it's a null entity, if it has a tombstone version and/or if both are valid. In most of the cases, the library is only interested in one or the other. However, there are some optimizations that rely on both of them being valid for an identifier.

**Comment by emsai** [comment]
I am also interested in more robust handling of the version numbers. I have started a general discussion related to this and proposed another approach here: https://github.com/skypjack/entt/discussions/875.
